File: packages/app/src/lib/session-export/sanitize.ts
import type { OpenCodeMessage } from "./types";

const MEDIA_PART_TYPES = new Set(["image", "video", "audio", "file"]);
const INLINE_PAYLOAD_KEYS = new Set(["base64", "binary", "bytes", "data"]);
const DATA_URL_RE = /^data:[^;]+;base64,/i;
const BASE64_LIKE_RE = /^[A-Za-z0-9+/=\s]+$/;
const INLINE_PAYLOAD_MIN_LENGTH = 256;

function isMediaPart(part: unknown): boolean {
  return (
    typeof part === "object" &&
    part !== null &&
    MEDIA_PART_TYPES.has(String((part as Record<string, unknown>).type))
  );
}

function isLargeInlinePayload(value: unknown): boolean {
  if (typeof value !== "string") {
    return false;
  }
  if (value.length < INLINE_PAYLOAD_MIN_LENGTH) {
    return false;
  }
  if (DATA_URL_RE.test(value)) {
    return true;
  }
  const compact = value.replace(/\s/g, "");
  return (
    compact.length >= INLINE_PAYLOAD_MIN_LENGTH && BASE64_LIKE_RE.test(compact)
  );
}

function sanitizeValue(value: unknown): unknown {
  if (Array.isArray(value)) {
    return sanitizeList(value);
  }
  if (typeof value === "object" && value !== null) {
    return sanitizeDict(value as Record<string, unknown>);
  }
  return value;
}

function sanitizeDict(data: Record<string, unknown>): Record<string, unknown> {
  const cleaned: Record<string, unknown> = {};
  for (const [key, value] of Object.entries(data)) {
    const lowerKey = key.toLowerCase();
    if (INLINE_PAYLOAD_KEYS.has(lowerKey) && isLargeInlinePayload(value)) {
      continue;
    }
    if (isLargeInlinePayload(value)) {
      continue;
    }
    cleaned[key] = sanitizeValue(value);
  }
  return cleaned;
}

function sanitizeList(items: unknown[]): unknown[] {
  const cleaned: unknown[] = [];
  for (const item of items) {
    if (isMediaPart(item)) {
      continue;
    }
    cleaned.push(sanitizeValue(item));
  }
  return cleaned;
}

function sanitizeMessage(message: OpenCodeMessage): OpenCodeMessage {
  const cleaned = sanitizeDict(message) as OpenCodeMessage;
  if (Array.isArray(message.parts)) {
    cleaned.parts = sanitizeList(message.parts) as Array<Record<string, unknown>>;
  }
  return cleaned;
}

export function sanitizeOpenCodeMessages(
  messages: OpenCodeMessage[],
): OpenCodeMessage[] {
  return messages.map(sanitizeMessage);
}

File: apps/desktop/src/session_export/sanitize.rs
use serde_json::{Map, Value};

const INLINE_PAYLOAD_MIN_LENGTH: usize = 256;
const INLINE_PAYLOAD_KEYS: [&str; 4] = ["base64", "binary", "bytes", "data"];
const MEDIA_PART_TYPES: [&str; 4] = ["image", "video", "audio", "file"];

fn is_media_part(part: &Value) -> bool {
    let Some(obj) = part.as_object() else {
        return false;
    };
    let Some(kind) = obj.get("type").and_then(Value::as_str) else {
        return false;
    };
    MEDIA_PART_TYPES.contains(&kind)
}

fn is_large_inline_payload(value: &Value) -> bool {
    let Some(raw) = value.as_str() else {
        return false;
    };
    if raw.len() < INLINE_PAYLOAD_MIN_LENGTH {
        return false;
    }
    if raw
        .get(..22)
        .map(|prefix| prefix.eq_ignore_ascii_case("data:") && raw.contains(";base64,"))
        .unwrap_or(false)
    {
        return true;
    }

    let compact = raw.chars().filter(|ch| !ch.is_whitespace());
    let mut len = 0usize;
    for ch in compact {
        len += 1;
        if !(ch.is_ascii_alphanumeric() || matches!(ch, '+' | '/' | '=')) {
            return false;
        }
    }
    len >= INLINE_PAYLOAD_MIN_LENGTH
}

fn sanitize_value(value: &Value) -> Value {
    match value {
        Value::Array(items) => sanitize_list(items),
        Value::Object(obj) => Value::Object(sanitize_dict(obj)),
        _ => value.clone(),
    }
}

fn sanitize_dict(data: &Map<String, Value>) -> Map<String, Value> {
    let mut cleaned = Map::new();
    for (key, value) in data {
        let lower = key.to_ascii_lowercase();
        if INLINE_PAYLOAD_KEYS.contains(&lower.as_str()) && is_large_inline_payload(value) {
            continue;
        }
        if is_large_inline_payload(value) {
            continue;
        }
        cleaned.insert(key.clone(), sanitize_value(value));
    }
    cleaned
}

fn sanitize_list(items: &[Value]) -> Value {
    let mut cleaned = Vec::with_capacity(items.len());
    for item in items {
        if is_media_part(item) {
            continue;
        }
        cleaned.push(sanitize_value(item));
    }
    Value::Array(cleaned)
}

fn sanitize_message(message: &Value) -> Value {
    let Some(obj) = message.as_object() else {
        return message.clone();
    };
    let mut cleaned = sanitize_dict(obj);
    if let Some(parts) = obj.get("parts").and_then(Value::as_array) {
        cleaned.insert("parts".to_string(), sanitize_list(parts));
    }
    Value::Object(cleaned)
}

pub fn sanitize_opencode_messages(messages: &[Value]) -> Vec<Value> {
    messages.iter().map(sanitize_message).collect()
}
